src/shell/notifications.test.js

```diff
@@ -0,0 +1,88 @@
+import { screen } from '@testing-library/react'
+import { within } from '@testing-library/dom'
+import userEvent from '@testing-library/user-event'
+
+import { endDay, getItemByName } from '../test-utils/ui'
+import { STORM_MESSAGE } from '../strings'
+import { farmhandStub } from '../test-utils/stubs/farmhandStub'
+import { saveFileStubFactory } from '../test-utils/stubs/saveFileStubFactory'
+
+describe('notifications', () => {
+  test('pending notifications are shown when a day starts', async () => {
+    jest.spyOn(Math, 'random').mockReturnValue(0)
+
+    await farmhandStub()
+    await endDay()
+
+    // STORM_MESSAGE message is a sample pending day notification that is known
+    // to show upon starting a new day when Math.random() is locked to 0.
+    await screen.findByText(STORM_MESSAGE)
+  })
+
+  test('notification is shown when recipe is learned', async () => {
+    const loadedState = saveFileStubFactory({
+      inventory: [{ id: 'carrot', quantity: 10 }],
+    })
+
+    await farmhandStub({
+      localforage: {
+        getItem: () => Promise.resolve(loadedState),
+        setItem: (_key, data) => Promise.resolve(data),
+      },
+    })
+
+    const carrotItem = await getItemByName('Carrot')
+    const carrotInput = within(carrotItem).getByDisplayValue('1')
+    userEvent.type(carrotInput, '10')
+    const carrotSellButton = within(carrotItem).getByText('Sell')
+    userEvent.click(carrotSellButton)
+    const notification = await screen.findByRole('alert')
+
+    within(notification).getByText('You learned a new recipe:', {
+      exact: false,
+    })
+    within(notification).getByText('Carrot Soup')
+  })
+
+  test('multiple notifications are shown when multiple recipes are learned', async () => {
+    const loadedState = saveFileStubFactory({
+      inventory: [
+        { id: 'carrot', quantity: 10 },
+        { id: 'corn', quantity: 6 },
+        { id: 'spinach', quantity: 30 },
+      ],
+    })
+
+    await farmhandStub({
+      localforage: {
+        getItem: () => Promise.resolve(loadedState),
+        setItem: (_key, data) => Promise.resolve(data),
+      },
+    })
+
+    const cornItem = await getItemByName('Corn')
+    const spinachItem = await getItemByName('Spinach')
+    const carrotItem = await getItemByName('Carrot')
+
+    const cornInput = within(cornItem).getByDisplayValue('1')
+    const spinachInput = within(spinachItem).getByDisplayValue('1')
+    const carrotInput = within(carrotItem).getByDisplayValue('1')
+
+    const cornSellButton = within(cornItem).getByText('Sell')
+    const spinachSellButton = within(spinachItem).getByText('Sell')
+    const carrotSellButton = within(carrotItem).getByText('Sell')
+
+    userEvent.type(cornInput, '6')
+    userEvent.click(cornSellButton)
+
+    userEvent.type(spinachInput, '30')
+    userEvent.click(spinachSellButton)
+
+    userEvent.type(carrotInput, '10')
+    userEvent.click(carrotSellButton)
```

Comment:
this ends up looking like a test that is checking to make sure recipe unlock logic is working rather than simply the notifications. maybe that was your goal, but it stands out to me as a bit overly complex and more of an integration rather than a unit test. it makes me wonder if there's a util that exists (or should exist) that can be mocked such as `getTodaysNotifications` instead?

Reply:
I will give that a read though I'm definitely not anti-integration test. I think the part that stood out to me as convoluted was that you need to understand recipe ingredients to understand this test, rather than just the fact that it's testing how multiple notifications work together. I'll give it some more thought on a fresh coffee though (and after reading that article, thanks for sharing!)

Reply:
a comment to me always says there's room for a code change to better explain what's going on, but i don't really want to be _too_ fussy over this. reading the comment though made me think maybe there's room for a method similar to what I called out but local to this spec. that way when looking at this test you can read it like:

```
unlockSummerSalad()
unlockCarrotSoup()

// expect some notifications...
```